### PR TITLE
query silliness

### PR DESCRIPTION
**Pull request by mind04**
### Short description
PowerDNS was  wasting a lot of queries while processing notifies
Use comboaddress for IPs (was strings)
Add 'supermaster' setting

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled and tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsbackend.hh [hunk 1]
```diff
@@ -86,6 +87,16 @@ struct DomainInfo
       return DomainInfo::Native;
   }
 
+  const bool isMaster(const ComboAddress& ip)
+  {
+    for( const auto& master: masters) {
+      const ComboAddress caMaster(master);
```
Comment (ahupowerdns) [inline pdns/dnsbackend.hh]: this needs to compare only IP addresses, and not also the port, I'm almost sure :-)

File: pdns/slavecommunicator.cc [hunk 2]
```diff
@@ -745,15 +745,7 @@ void CommunicatorClass::slaveRefresh(PacketHandler *P)
...
-    int res;
-    res=P->trySuperMasterSynchronous(&dp, tsigkeyname);
-    if(res>=0) {
-      DNSPacket *r=dp.replyPacket();
-      r->setRcode(res);
-      r->setOpcode(Opcode::Notify);
-      N->send(r);
-      delete r;
-    }
+    P->trySuperMasterSynchronous(&dp, tsigkeyname); // FIXME could use some error loging
```
Comment (ahupowerdns) [inline pdns/slavecommunicator.cc]: was there a specific reason for removing the error packet generation? or can it no longer fail?

File: modules/gmysqlbackend/gmysqlbackend.cc [hunk 3]
```diff
@@ -94,8 +94,6 @@ class gMySQLFactory : public BackendFactory
     declare(suffix, "remove-empty-non-terminals-from-zone-query", "remove all empty non-terminals from zone", "delete from records where domain_id=? and type is null");
     declare(suffix, "delete-empty-non-terminal-query", "delete empty non-terminal from zone", "delete from records where domain_id=? and name=? and type is null");
 
-    declare(suffix,"master-zone-query","Data", "select master from domains where name=? and type='SLAVE'");
```
Comment (ahupowerdns) [inline modules/gmysqlbackend/gmysqlbackend.cc]: <3 cleanup!

File: pdns/dnsbackend.hh [hunk 4]
```diff
@@ -51,7 +52,7 @@ struct DomainInfo
   DNSName zone;
   time_t last_check;
   string account;
-  vector<string> masters;
+  vector<string> masters; // FIXME use ComboAddress
```
Comment (ahupowerdns) [inline pdns/dnsbackend.hh]: I can contribute that cleanup if you want, and commit it to this PR?

Comment (mind04) [reply pdns/dnsbackend.hh]: Would be awesome.

File: pdns/dnsbackend.hh (hunk 1 above)
Comment (mind04) [reply pdns/dnsbackend.hh]: I almost sure you are right  ;) Fixing...

File: pdns/slavecommunicator.cc (hunk 2 above)
Comment (mind04) [reply pdns/slavecommunicator.cc]: this was responsible for the spurious answer (log is in the commit message). After removal of this piece of code the error is gone. There is no way to let the master know at this point it is failing.
A nice error in the log is all we can do here.

**Comment by pieterlexis** [comment]
Judging by the title update, this is ready for another review?

**Comment by mind04** [comment]
Yes, ready for review.

**Pull request merged**